src/client.js

```diff
@@ -39,100 +41,115 @@ class Client extends EventEmitter {
-      this.emit('error', e)
-    })
+    if (!this.customCommunication) {
+      this.serializer = createSerializer({ isServer: this.isServer, version: this.version, state, customPackets: this.customPackets })
+      this.deserializer = createDeserializer({
+        isServer: this.isServer,
+        version: this.version,
+        state,
+        packetsToParse:
+          this.packetsToParse,
+        customPackets: this.customPackets,
+        noErrorLogging: this.hideErrors
+      })
+
+      this.splitter.recognizeLegacyPing = state === states.HANDSHAKING
+
+      this.serializer.on('error', (e) => {
+        let parts
+        if (e.field) {
+          parts = e.field.split('.')
+          parts.shift()
+        } else { parts = [] }
+        const serializerDirection = !this.isServer ? 'toServer' : 'toClient'
+        e.field = [this.protocolState, serializerDirection].concat(parts).join('.')
+        e.message = `Serialization error for ${e.field} : ${e.message}`
+        if (!this.compressor) { this.serializer.pipe(this.framer) } else { this.serializer.pipe(this.compressor) }
+        this.emit('error', e)
+      })
+
+      this.deserializer.on('error', (e) => {
+        let parts
+        if (e.field) {
+          parts = e.field.split('.')
+          parts.shift()
+        } else { parts = [] }
+        const deserializerDirection = this.isServer ? 'toServer' : 'toClient'
+        e.field = [this.protocolState, deserializerDirection].concat(parts).join('.')
+        e.message = `Deserialization error for ${e.field} : ${e.message}`
+        if (!this.compressor) { this.splitter.pipe(this.deserializer) } else { this.decompressor.pipe(this.deserializer) }
+        this.emit('error', e)
+      })
+    }
+
     this._mcBundle = []
     const emitPacket = (parsed) => {
       this.emit('packet', parsed.data, parsed.metadata, parsed.buffer, parsed.fullBuffer)
       this.emit(parsed.metadata.name, parsed.data, parsed.metadata)
       this.emit('raw.' + parsed.metadata.name, parsed.buffer, parsed.metadata)
       this.emit('raw', parsed.buffer, parsed.metadata)
     }
-    this.deserializer.on('data', (parsed) => {
-      parsed.metadata.name = parsed.data.name
-      parsed.data = parsed.data.params
-      parsed.metadata.state = state
-      debug('read packet ' + state + '.' + parsed.metadata.name)
-      if (debug.enabled) {
-        const s = JSON.stringify(parsed.data, null, 2)
-        debug(s && s.length > 10000 ? parsed.data : s)
-      }
-      if (parsed.metadata.name === 'bundle_delimiter') {
-        if (this._mcBundle.length) { // End bundle
-          this._mcBundle.forEach(emitPacket)
-          emitPacket(parsed)
-          this._mcBundle = []
-        } else { // Start bundle
+
+    if (this.customCommunication) {
+      this.customCommunication.receiverSetup.call(this, (/** @type {{name, params, state?}} */parsed) => {
```

Comment:
The fact customCommunication should have a receiverSetup property in it with a call function should be documented